Post Editor:  Add image selector component for multiple post images or post meta

joshuatf opened the pull request:
This PR adds a reusable component that allows for multiple images to be selected as featured images or for various post meta options using the media modal for image selection and uploads.

The existing editor-featured-image does this well for a single image but does not allow multiple images to be selected and also handles data persistence directly within the component.  

This new component will be in many ways similar to the featured image, but will allow multiple images and also handle data changes within a prop to allow for easy extensibility.

Initially, this is planned for product image selection using the media modal (related to issue #21302 ) as well as product variant image selection.  It could also be used in place of the existing featured image component to further reduce code duplication.

This branch depends on #26014

timmyc commented on client/post-editor/editor-image-selector/dropzone-icon.jsx:
```diff
@@ -0,0 +1,24 @@
+
+export default () => (
+	<svg width="55" height="42" viewBox="0 0 55 42" xmlns="http://www.w3.org/2000/svg">
+		<title>Image Selector</title>
```
This string needs to be passed through `translate`.

joshuatf replied:
Updated!

timmyc commented on client/post-editor/editor-image-selector/dropzone.jsx:
```diff
@@ -0,0 +1,101 @@
+		);
+	}
+}
+
```
Could we add some `propType` declarations here?

joshuatf replied:
Absolutely.  Added the declarations here.

timmyc commented on client/post-editor/editor-image-selector/dropzone.jsx:
```diff
@@ -0,0 +1,101 @@
+
+				// Successful image upload.
+				if ( media ) {
+					this.props.recordTracksEvent( 'calypso_editor_image_selector_upload', {
```
Since this component might be used elsewhere, wondering about the naming of this track event.

joshuatf replied:
You're right.  Since it's going to be used elsewhere I think the best way to handle this is to remove it entirely and handle it in the props when the dropped image is passed back.

timmyc commented on client/post-editor/editor-image-selector/index.jsx:
```diff
@@ -0,0 +1,163 @@
+		translate: PropTypes.func,
+		onImageChange: PropTypes.func,
+		onImageSelected: PropTypes.func,
+		onRemoveImage: PropTypes.func,
```
missing `siteId` here.

joshuatf replied:
Added!

timmyc commented on client/post-editor/editor-image-selector/index.jsx:
```diff
@@ -0,0 +1,163 @@
+		hasDropZone: false,
+		isDropZoneVisible: false,
+		maxWidth: 450,
+		onImageSelected: noop,
```
instead of having 3 distinct functions for selected, change, remove - why not just have one onchange event?

timmyc replied:
Disregard - I see it is the underlying components that declare these :)

timmyc commented on client/post-editor/editor-image-selector/index.jsx:
```diff
@@ -0,0 +1,163 @@
+					labels={ { confirm: multiple ? translate( 'Set images' ) : translate( 'Set image' ) } }
+					enabledFilters={ [ 'images' ] }
+					galleryViewEnabled={ false }
+					{ ...! multiple && { single: true } }
```
Could this line be simplified to something like `single: ! multiple`? Especially if we set the default prop for multiple to false?

timmyc commented on client/post-editor/editor-image-selector/preview.jsx:
```diff
@@ -0,0 +1,223 @@
+import { url } from 'lib/media/utils';
+
+export class EditorImageSelectorPreview extends React.Component {
+	static displayName = 'EditorImageSelectorPreview';
```
I don't think this is required anymore, also why not just import `Component` directly and extend it here?

joshuatf replied:
Updated both.

timmyc commented on client/post-editor/editor-image-selector/preview.jsx:
```diff
@@ -0,0 +1,223 @@
+					return media;
+				} )
+				.filter( function( e ) {
+					return e;
```
So just confirming, this is filtering out images that are actively being fetched, and as such falsey from above?

timmyc commented on client/post-editor/editor-image-selector/preview.jsx:
```diff
@@ -0,0 +1,223 @@
+
+		this.setState( { images }, () => {
+			if ( 'function' === typeof callback ) {
+				callback();
```
Can you explain a bit why this particular callback on the setState is needed?

joshuatf replied:
I had handled this a bit incorrectly and udpated here - https://github.com/Automattic/wp-calypso/pull/26297/commits/cc8ed5029c0f70e3e30e9d515192a4a630969aa6

Images shouldn't be fetched every time we update the image state, only in the callback which checks if the props and image state are different and then fetches the images if not.  

We actually do need to run the updateImageState regularly to show transient images and fetched images asynchronously.

timmyc commented on client/post-editor/editor-image-selector/preview.jsx:
```diff
@@ -0,0 +1,223 @@
+			<div className={ classes } key={ id }>
+				<Button
+					className="editor-image-selector__image"
+					onClick={ this.props.onImageClick }
```
could add `onImageClick` to line 135 to eliminate repeating `this.props` here.

joshuatf replied:
Destructured this and a few other props here.

timmyc commented on client/post-editor/editor-image-selector/test/index.jsx:
```diff
@@ -0,0 +1,154 @@
+jest.mock( 'event', () => {}, { virtual: true } );
+jest.mock( 'lib/media/store', () => ( {
+	dispatchToken: require( 'dispatcher' ).register( () => {} ),
+	get: ( siteId, itemId ) => require( './fixtures' ).DUMMY_MEDIA[ itemId ],
```
Nice!

joshuatf commented:
@timmyc Thanks for all the great feedback!  I'll get back to you on all of the individual points by tomorrow.

Regarding the location, I went back and forth trying to decide whether this should be under post-editor or components.  My reasoning for picking post-editor was because I assumed it would be mostly used under posts (including products if we can classify them as a type of post) and that's where the existing featured image component resides.

It's certainly feasible that this is used outside of posts entirely though, in which case, components may be a better option.  It sounds like you're leaning that way- thoughts?

a8ci18n commented on client/components/image-selector/preview.jsx:
```diff
@@ -0,0 +1,220 @@
+			compact,
+		} );
+
+		const addString = multiple ? translate( 'Add images' ) : translate( 'Add image' );
```
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

joshuatf commented:
@timmyc I moved this into the components instead of post-editor and added a README as well an example implementation.

Let me know your thoughts on whether this fits better with this component's MO.  Thanks again 😄

timmyc commented on client/components/image-selector/docs/example.jsx:
```diff
@@ -0,0 +1,68 @@
+/** @format */
```
This is great stuff and a very thorough README - could you get it added in to the `devdocs` area so the example can be seen in action there?

joshuatf replied:
Sure!  I actually wanted to do this, but I have a quick question here.  Since this component also relies on the media modal styles, what's the accepted way to import these files?  Toss it into `devdocs.scss` or is there a better way to get this dependency imported?

joshuatf replied:
@timmyc Cool.  Just added to the devdocs/blocks section for testing.  I moved the component to blocks as this seems a bit more appropriate since it's not a primitive UI component.

https://github.com/Automattic/wp-calypso/pull/26297/commits/a920d11283ce1152fcefd88bbaaff27ebf496539

joshuatf commented on client/post-editor/editor-image-selector/preview.jsx:
```diff
@@ -56,32 +56,29 @@ export class EditorImageSelectorPreview extends Component {
 		this.updateImageState( () => {
-			if ( this.state.images ) {
+			const { itemIds, siteId } = this.props;
+			if ( isEqual( this.state.images.map( image => image.ID ), itemIds ) ) {
```
It's possible in this callback that if the images were reordered, this would return falsey, but I've added another check to determine whether or not the item is already in the MediaStore so we don't fetch again if so.

https://github.com/Automattic/wp-calypso/blob/2af049b2796faea2d23fd6439ba85dadd7e8eb82/client/components/image-selector/preview.jsx#L67-L70

We could sort the arrays and compare equality to return early here, but we would still want to check the store for these files anyway to avoid duplicate fetches since the media library and dropzone handle some fetches for us.

joshuatf commented on client/post-editor/editor-image-selector/preview.jsx:
```diff
@@ -94,8 +91,14 @@ export class EditorImageSelectorPreview extends Component {
+				onImageChange &&
+				images &&
+				images.length === itemIds.length &&
+				! isEqual( imageIds, itemIds )
```
If the order is different here, we actually would want this to trigger the callback so we can send the reordered data back to the parent component.

a8ci18n commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,223 @@
+			compact,
+		} );
+
+		const addString = multiple ? translate( 'Add images' ) : translate( 'Add image' );
```
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

timmyc commented on client/blocks/image-selector/docs/example.jsx:
```diff
@@ -0,0 +1,89 @@
+                <ImageSelector
+					site={ this.props.site }
+					siteId={ this.props.siteId }
+                    imageIds={ imageIds }
```
tabbing is a little odd on some of the props here.

timmyc commented on client/blocks/image-selector/docs/example.jsx:
```diff
@@ -0,0 +1,89 @@
+                    onAddImage={ this.addImage }
+                    multiple={ true }
+					showEditIcon={ true }
+					translate={ translate }
```
Does translate need to be passed as a prop in this way or can `ImageSelector` just use the `localize` [HOC](https://github.com/automattic/i18n-calypso#localize) directly?

timmyc replied:
Looks like it already does, so this prop isn't needed here.

timmyc commented on client/blocks/image-selector/index.jsx:
```diff
@@ -0,0 +1,173 @@
+		const { imageIds } = this.props;
+		const classes = classnames( 'image-selector', {
+			'is-assigned': !! imageIds && imageIds.length,
+			'has-active-drop-zone': this.props.hasDropZone && this.props.isDropZoneVisible,
```
Minor but you could add `hasDropZone` and `isDropZoneVisible` to the deconstructions above.

timmyc commented on client/blocks/image-selector/index.jsx:
```diff
@@ -0,0 +1,173 @@
+				{ this.props.hasDropZone && (
+					<ImageSelectorDropZone
+						onDroppedImage={ this.addImage }
+						site={ this.props.site }
```
likewise on `site` and `siteId`

timmyc commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,223 @@
+		const featuredImage = ( images && images.shift() ) || null;
+
+		return (
+			<div className="image-selector__images-wrapper">
```
Do you think having a `classname` prop would be of use here so when this is used in other components the styling could be adjusted accordingly?

joshuatf commented:
@timmyc I did run into the same issue with uploading media, but it does seem to be part of the master branch as well.  It's probably related to Chrome's restrictions against sending data across an insecure network.  

I believe adding a self-signed certificate and enabling chrome://flags/#allow-insecure-localhost would work, but a quick workaround is to test with Safari or Firefox.

timmyc commented:
Ah that makes sense, I'll give it a test in Firefox too - thanks @joshuatf

timmyc commented:
Upload / transients are working great in my testing. I'm going to get one more person to check out the PR.

a8ci18n commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,227 @@
+			compact,
+		} );
+
+		const addString = multiple ? translate( 'Add images' ) : translate( 'Add image' );
```
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

joshuatf commented on client/blocks/image-selector/dropzone.jsx:
```diff
@@ -0,0 +1,112 @@
+
+export default connect(
+	( state, ownProps ) => {
+		const { site, siteId } = ownProps;
```
You're right; removed site so we can just pass in siteId for the devdocs.

joshuatf commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,227 @@
+ */
+import Button from 'components/button';
+import ImagePreloader from 'components/image-preloader';
+import { localize } from 'i18n-calypso';
```
Gotcha- thanks!

joshuatf commented on client/blocks/image-selector/docs/example.jsx:
```diff
@@ -0,0 +1,89 @@
+					onImageChange={ this.changeImages }
+					onRemoveImage={ this.removeImage }
+					onAddImage={ this.addImage }
+					multiple={ true }
```
Updated.

joshuatf commented:
@justinshreve Thanks so much for reviewing!  I updated the transients a bit to act as placeholders so we can avoid that flash.  Could you let me know if things look okay now?

a8ci18n commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,236 @@
+			compact,
+		} );
+
+		const addString = multiple ? translate( 'Add images' ) : translate( 'Add image' );
```
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

nb replied:
@joshuatf could you look at those @a8ci18n warnings and act on them somehow?

joshuatf replied:
@nb Sure, but I'm a bit unsure on how to act on this.  Is there anything else needed aside from 👍or 👎?

joshuatf replied:
Ah- gotcha!  Completely missed the capital 'i' at first glance 😅

joshuatf replied:
Spoke too soon; still getting that warning.  Do we need to include the same context to suppress the warning?

nb replied:
Well, the end goal is to reuse the translation, not just to suppress the warning :) And you're right, we need the context in order to reuse it. This is unfortunate, ideally this wouldn't have been context, but just a translator comment since contexts are used for disambiguation. The string [comes from WordPress core](https://core.trac.wordpress.org/browser/trunk/src/wp-includes/widgets/class-wp-widget-media-gallery.php?rev=42343#L35).

nb commented on client/blocks/image-selector/README.md:
```diff
@@ -0,0 +1,61 @@
+Name | Type | Default | Description
+--- | --- | --- | ---
+`className` | `string` | `` | String used for the class of outer ImageSelector wrapper
+`compact` | `bool` | `0` | Determines whether or not the upload picker is compact or full size
```
Do you envision us using the `compact` prop anywhere?

joshuatf replied:
Definitely.  I've already used it in the branch that covers product variations as we always want this uploader compact since the product variation rows are pretty small.

joshuatf replied:
Sure thing.  Let me rebase and I'll push in a few minutes.

joshuatf replied:
@nb Here's the product image branch where you can see this use case in action https://github.com/Automattic/wp-calypso/pull/26703

nb commented on client/blocks/image-selector/README.md:
```diff
@@ -0,0 +1,61 @@
+`maxWidth` | `number` | `450` | Maximum width to be used for a preview image
+`onAddImage` | `func` | `noop` | Passes a single transient image object after an image is dropped on the DropZone
+`onImageChange` | `func` | `noop` | Passes the entire array of currently selected image objects when transitioning from a transient to permanent image
+`onImageSelected`* | `func` | `noop` | Passes an array of selected image objects when set from the Media Modal
```
Does the `*` mean that the prop is required? If yes, shouldn't we also mark it as required in the `propTypes`, too?

joshuatf replied:
Absolutely, good call.  My intention originally was to not require any of these props, but I can't come up with any use cases currently where we wouldn't want to handle images when passed back.

nb commented on client/blocks/image-selector/README.md:
```diff
@@ -0,0 +1,61 @@
+
+### Persistence and Tracking
+
+The Image Selector component intentionally does not persist any information.  Handling image state, saving, and tracking should be done when image changes are passed back:
```
I really like this design decision, it makes it reusing it a lot easier.

nb commented on client/blocks/image-selector/README.md:
```diff
@@ -0,0 +1,61 @@
+--- | --- | --- | ---
+`className` | `string` | `` | String used for the class of outer ImageSelector wrapper
+`compact` | `bool` | `0` | Determines whether or not the upload picker is compact or full size
+`hasDropZone` | `bool` | `0` | Enables dropzone for drag-and-drop images
```
In what cases one would like to disable the drop zone?

joshuatf replied:
I think in a compact area (e.g., product variations), the drop zone may not be ideal UX since the drop zone space is pretty limited.  What do you think?

joshuatf replied:
Other than not having the drop zone itself, I don't believe it affects the component in any way.  Items are still selectable and uploadable via the MediaModal.

nb replied:
I don't think I was clear – what if we eliminated the `hasDropZone` prop and made it depend on the `compact` prop?

```js
const hasDropZone = ! this.props.compact;
```

joshuatf replied:
Ah, I see.  That's a good idea.  I'll get this added.

joshuatf replied:
Actually on second thought, the compact prop is used to determine the upload picker size, so doing this would disable the drop zone after the first image is selected in the example.

We set the value of compact to true when more than one image id is passed to create a compact uploader for non-featured images.

nb commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,236 @@
+				return;
+			}
+
+			defer( () => {
```
Why do we need the `defer` here?

joshuatf replied:
I believe this and the other defer you mentioned below are from the initial commit before image fetching was moved to the setState callback.  

They were needed to make sure the images were fetched correctly on mount, but I'm unable to reproduce right now so I think they can be safely removed.

nb replied:
👍

nb commented on client/blocks/image-selector/dropzone.jsx:
```diff
@@ -0,0 +1,109 @@
+ */
+import DropZone from 'components/drop-zone';
+import MediaActions from 'lib/media/actions';
+import MediaStore from 'lib/media/store';
```
Have you looked at using the newer methods to access the media data, like `redux` or the data layer? I think #26168 has some good context on this.

joshuatf replied:
I did look into this, but I'm a bit confused by how it's supposed to be implemented (or if parts are ready to be implemented).

The QueryMedia component only accepts a single `mediaId` as a prop.  It seems like it may be able to handle multiple images if passed through `query`, but I can't find any other reference to this or how to form the query.  Or is the best method here to form individual queries for each of the images?

For the new data layer, it looks like there's currently not a way to handle uploading files, so I'm not sure if we can replace the current Flux-based MediaActions implementation.

nb replied:
We will be replacing all Flux implementations at some point and I’d rather us not introduce new store uses. Pinging @Automattic/team-calypso to see if there are obvious next steps we're missing.

joshuatf replied:
@nb Understood.  I'll dig around and see if I can't find more information on this while we're waiting for a response.

timmyc replied:
I chatted with @jsnajdr about the migration of the media store from flux to redux today. It is slated to be updated ( p3fqKv-6uU-p2 ) but no ETA and it is one of the more complex stores. So I guess the question here is are we okay adding in additional usage of the current flux stores which will have to be migrated during the update?

joshuatf replied:
@timmyc When I chatted with @nb I think we got the okay to go ahead and merge since it looked like the Redux implementation for media actions needs more time but agreed let's wait for @nb to double confirm that.

We can always hold off on this merge and rebase once the Redux stores are ready.

nb replied:
Let’s merge this if you don't have any blockers, @timmyc – I’d rather we get a better image selector to users sooner than wait on reduxification.

nb commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,236 @@
+				callback();
+			}
+		} );
+		defer( () => {
```
Why do we need `defer` here?

joshuatf replied:
(Same as above).  Removed.

nb commented:
Thanks for working on this @joshuatf, it's a great effort 💯 It's great to see the much simple extracted interface of the image selector component. I left some comments in the code, but none of them are blockers.

A small bug (or a feature) was that the component allowed dragging and dropping an already selected image, which duplicated it in the media library.

One thing I am missing is a quick sketch for how we are going to use the component in Calypso – the specific use-cases should inform the interface of the component and it's hard for me to judge which of the props make sense without seeing how it will be used in action. Thanks!

joshuatf commented:
@nb My pleasure- thanks for reviewing!

I did notice the bug where selected images could be dragged in Safari.  Should be fixed by https://github.com/Automattic/wp-calypso/pull/26297/commits/4201ffd7355b0ca403095a21f6de80242e2e67d4

## Use cases

Here are a few use cases for the ImageSelector component with images attached for those already planned and added in an upcoming PR.

### Product images

Product images were the primary use case for this component.  We want to be able to add multiple images while showing a singular 'featured image.'

The compact prop allows us to initially show a larger upload picker followed by a smaller one after the first image is set.

<img width="1050" alt="screen shot 2018-08-15 at 5 38 06 pm" src="https://user-images.githubusercontent.com/10561050/44142245-3e8c58ae-a0b2-11e8-9d56-044dd0b54fac.png">

###  Post Featured Images

This component could also be used to replace the existing featured image component.  We would probably want to add an extra prop for more specific messages if this is implemented (e.g., instead of "Add Image" we could pass in "Set Featured Image."

### Product variations

Since the area for variations is small (a single table cell), we want to show a compact uploader intiially and limit to a single image (i.e., not passing the multiple prop).  I think it may also be best to not show the drop zone here with this small of a space, but interested in hearing your opinion.

<img width="420" alt="screen shot 2018-08-15 at 5 38 58 pm" src="https://user-images.githubusercontent.com/10561050/44142228-352935fc-a0b2-11e8-9e12-4bcb922d6905.png">

### Category images

Similar use case to the product variations.

### Other post meta fields

I don't think they currently exist on Calypso, but eventually, this could also be used for meta fields in custom post types if they are ever integrated.

a8ci18n commented on client/blocks/image-selector/preview.jsx:
```diff
@@ -0,0 +1,237 @@
+			compact,
+		} );
+
+		const addString = multiple ? translate( 'Add Images' ) : translate( 'Add Image' );
```
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n replied:
Hi! I've found a possible matching string that has already been translated **57** times:
`translate( 'Add Images', { context: 'label for button in the gallery widget; should not be longer than ~13 characters long'} )` *ES Score: 11*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).
